### DISCUSSION TITLE
How does DMPD deal with different local neighbour count?

### DISCUSSION THREAD

**Discussion by mearcstapa-gqz** [deepmodeling/deepmd-kit#670] [General]
Since the neural network need to take a fix-sized input, how does DMPD get the input to the MLP when the local neighbour count is different? The DMPD paper only mentioned that it uses sorted input data for an atom (such that it preserves symmetry), but for a different atom, the input length (neighbour count) could be different. Could you point me to the code where the model handles this?

**Comment by njzjz** [comment]
The size is set by users. https://github.com/deepmodeling/deepmd-kit/blob/f765b483f04116f53e9a9e44f6023a1bcea121cc/examples/water/train/water_se_a.json#L7

**Reply by caolz** [reply]
> The size is set by users.
> 
> https://github.com/deepmodeling/deepmd-kit/blob/f765b483f04116f53e9a9e44f6023a1bcea121cc/examples/water/train/water_se_a.json#L7

I know that when you train the combustion potential function, you treat compounds as clusters, so does the setting of "sel" need to consider periodic boundary conditions? How can I set "sel" without considering the periodic boundary conditions?

**Reply by mearcstapa-gqz** [reply]
Sorry I don't get it. Say there's an O atom with 20 neighbouring atom within cut-off radius and there's another O atom with 7 neighbouring atom within cut-off radius. How do model get a fixed sized input?

**Reply by njzjz** [reply]
> Sorry I don't get it. Say there's an O atom with 20 neighbouring atom within cut-off radius and there's another O atom with 7 neighbouring atom within cut-off radius. How do model get a fixed sized input?

For example, if you set to `[46， 92]`, the input size will be always 138 (46 for O neighbors and 92 for H neighbors). If neighboring atoms are not enough, the size will be expanded to what you set to by filling with zero.

**Reply by njzjz** [reply]
> > The size is set by users.
> 
> > 
> 
> > https://github.com/deepmodeling/deepmd-kit/blob/f765b483f04116f53e9a9e44f6023a1bcea121cc/examples/water/train/water_se_a.json#L7
> 
> 
> 
> I know that when you train the combustion potential function, you treat compounds as clusters, so does the setting of "sel" need to consider periodic boundary conditions? How can I set "sel" without considering the periodic boundary conditions?

I don't think there's any difference between the PBC and the non-PBC. You only need to consider neighbors within the cutoff radius, for both training and MD simulations.

**Reply by mearcstapa-gqz** [reply]
Thanks! To make sure, if an atom has more than 46 O neighbors in its cut-off radius, it will only consider the closest 46, right?

And,
Do deepmd support small organic molecules? Say,  molecules with ~40 atoms in C,H,O,N,F,P,Si,S,Cl,Br,I.

Thanks!!

**Reply by njzjz** [reply]
> Thanks! To make sure, if an atom has more than 46 O neighbors in its cut-off radius, it will only consider the closest 46, right?

Yes, but it's not suggested. It will lead to discontinuous potential energy surface. It's better to set `sel` larger than the number of neighbours.

> Do deepmd support small organic molecules?

Yes.
